### PR TITLE
Support manual instruction type override relations

### PR DESCRIPTION
**Pull request by danpat**
This is an experimental PR to tinker with the idea of data-driven instruction generation.

Rationale: sometimes, you want a specific instruction at a particular location, and the heuristics we're using just aren't fancy enough to make the right decision, or the OSM tagging scheme is incomplete for the structure of the map at that location and doesn't supply enough information to make a good decision about which instruction to issue.

This PR adds support for a new, totally made up relation we're calling a `maneuver relation`.  They look like this:

```
<relation>
  <tag k="type" v="maneuver"/>
  <member type="way" ref="1234" role="from"/>
  <member type="way" ref="5678" role="to"/>
  <member type="node" ref="9999" role="via"/>
  <tag k="maneuver" v="turn"/>
  <tag k="direction" v="slight_right"/>
</relation>
```

The basic process is that after a route is found, we scan along the path and find `maneuver` relations that match (they pass through `from`, `via`, and `to` in that order).  If a match is found, the `type` value for the `via` point is replaced with the value from the relation (same for the `direction`).

We will come up with a rough mapping of reasonable-looking `type` values to OSRM turn types - the intent is for the OSM data to be relatively routing-engine agnostic.  The intent of this relation type is to be a hint to the navigation engine in locations where heuristics have trouble.

## TODO:

  - [x] Complete the `RestrictionCompressor` work so that it also updates the ManeuverOverride from/to IDs @karenzshea
  - [x] Implement the lookup to find matching ManeuverOverrides in the EdgeBasedGraphFactory (currently stubbed out)
  - [x] Serialize the `ManeuverOverride` list at the end of extraction @danpat
  - [x] Fully implement the `facade.GetOverridesThatStartAt` method
  - [ ] Extend this so that from/to ways don't need to be directly connected **

### Multi-via-way support
  - [x] Fix relation compressor for multi-nodes
  - [x] Verify which nodes are needed `
  - [x] File saving for multi-nodes
  - [ ] Edge case test scenarios

** Do this bit after we have directly-connected overrides working

/cc @karenzshea 

## Tasklist
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [ ] add regression / cucumber cases (see docs/testing.md)
 - [ ] review
 - [ ] adjust for comments

### CODE REVIEW

File: src/extractor/extraction_containers.cpp [hunk 1]
```diff
@@ -795,6 +795,41 @@ void ExtractionContainers::PrepareManeuverOverrides()
...
+        } else if (turn_string == "fork") {
+            return guidance::TurnType::Fork;
+        } else if (turn_string == "suppress") {
+            return guidance::TurnType::Suppressed;
+        } else {
+            return guidance::TurnType::Invalid;
+        }
+    };
+
+    const auto string_to_turn_direction = [](const std::string &direction_string) {
```
Comment (karenzshea) [inline src/extractor/extraction_containers.cpp]: I think here we will also need to pass in the `turn_string` to be able to set a `uturn` direction modifier

**Comment by karenzshea** [comment]
@danpat Do you have some unpushed additions to the IO operations for reading/writing maneuver overrides? Locally and on every travis build I'm seeing a 
```
/home/travis/build/Project-OSRM/osrm-backend/include/extractor/serialization.hpp:339:39:   required from here
/home/travis/build/Project-OSRM/osrm-backend/include/storage/io.hpp:101:9: error: static assertion failed: bytewise reading requires trivially copyable type
```
error, where it looks like we're missing the `read` methods for the `ManeuverOverride` type.

File: include/engine/api/route_api.hpp [hunk 2]
```diff
@@ -139,6 +164,150 @@ class RouteAPI : public BaseAPI
...
+
+                        // We got a match, update using the instruction_node
+                        if (search_iter == maneuver_relation.node_sequence.end())
+                        {
+                            std::cout << "Node sequence matched, looking for the step "
+                                      << "that has the via node" << std::endl;
+                            const auto via_node_coords = BaseAPI::facade.GetCoordinateOfNode(
+                                maneuver_relation.instruction_node);
+                            // Find the step that has the instruction_node at the intersection point
+                            auto step_to_update = std::find_if(
```
[the remark this answers is not in the document]
Comment (danpat) [reply include/engine/api/route_api.hpp]: 🤔 This is a good idea and would definitely speed things up.  We're also being wasteful here by checking every coordinate - only intersection coordinates should get checked.

File: include/engine/datafacade/contiguous_internalmem_datafacade.hpp [hunk 3]
```diff
@@ -890,6 +910,40 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
...
+    std::vector<extractor::ManeuverOverride>
+    GetOverridesThatStartAt(const NodeID edge_based_node_id) const override final
+    {
+        std::vector<extractor::ManeuverOverride> results;
+
+        bool found = false;
+        // Copy all the overrides into the results array.
+        // m_maneuver_overrides is sorted by node_sequence.front(),
+        // so we can stop searching after we get to a node > edge_based_node_id
+        for (const auto & override : m_maneuver_overrides)
```
[the remark this answers is not in the document]
Comment (danpat) [reply include/engine/datafacade/contiguous_internalmem_datafacade.hpp]: Good catch - yeah, it would be faster, `std::equal_range` has sub-linear performance if the thing being searched is random-access (as opposed to just forward iterable).

The way we're comparing a member variable (`override.start_node`) doesn't fit with the `std::equal_range` pattern though, I'll see if I can re-jigger it to work.

File: include/extractor/maneuver_override.hpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        turn_sequence; // initially the internal node-based-node ID of the node
+    // before the turn, then later, the edge_based_node_id of the turn
+    NodeID instruction_node; // node-based node ID
+    guidance::TurnType::Enum override_type;
+    guidance::DirectionModifier::Enum direction;
+
+    // check if all parts of the restriction reference an actual node
+    bool Valid() const
+    {
+        return turn_sequence.size() >= 2 || std::none_of(turn_sequence.begin(),
```
[the remark this answers is not in the document]
Comment (karenzshea) [reply include/extractor/maneuver_override.hpp]: @danpat I'm still unclear on whether or not this check is the right one. Should the logic here be inverted?
